test/test_complex.py

```diff
@@ -0,0 +1,211 @@
+def test_conj(self):
+	z1 = ComplexValue(1+2j)
+	z2 = ComplexValue(1-2j)
+
+	assert z1 == Conj(z2)
+	assert z2 == Conj(z1)
+
+
+def test_real(self):
+	z0 = Zero()
+	z1 = as_ufl(1.0)
+	z2 = ComplexValue(1j)
+	z3 = ComplexValue(1+1j)
+
+	assert Real(z1) == z1
+	assert Real(z3) == z1
+	assert Real(z2) == z0
+
+
+def test_imag(self):
+	z0 = Zero()
+	z1 = as_ufl(1.0)
+	z2 = as_ufl(1j)
+	z3 = ComplexValue(1+1j)
+
+	assert Imag(z2) == z1
+	assert Imag(z3) == z1
+	assert Imag(z1) == z0
+
+
+def test_compute_form_adjoint(self):
+	cell = triangle
+	element = FiniteElement('Lagrange', cell, 1)
+
+	u = TrialFunction(element)
+	v = TestFunction(element)
+
+	a = inner(grad(u), grad(v))*dx
+
+	assert compute_form_adjoint(a) == conj(inner(grad(v), grad(u)))*dx
+
+
+def test_complex_algebra(self):
+	z1 = ComplexValue(1j)
+	z2 = ComplexValue(1+1j)
+
+	# Remember that ufl.algebra functions return ComplexValues, but ufl.mathfunctions return complex Python scalar
+	# Any operations with a ComplexValue and a complex Python scalar promote to ComplexValue
+	assert z1*z2 == ComplexValue(-1+1j)
+	assert z2/z1 == ComplexValue(1-1j)
+	assert pow(z2, z1) == ComplexValue((1+1j)**1j)
+	assert sqrt(z2)*as_ufl(1) == ComplexValue(cmath.sqrt(1+1j))
+	assert ((sin(z2) + cosh(z2) - atan(z2))*z1) == ComplexValue((cmath.sin(1+1j) + cmath.cosh(1+1j) - cmath.atan(1+1j))*1j)
+	assert (abs(z2) - ln(z2))/exp(z1) == ComplexValue((abs(1+1j) - cmath.log(1+1j))/cmath.exp(1j))
+
+
+def test_automatic_simplification(self):
+	cell = triangle
+	element = FiniteElement("Lagrange", cell, 1)
+
+	v = TestFunction(element)
+	u = TrialFunction(element)
+
+	assert inner(u, v) == u*conj(v)
+	assert dot(u, v) == u*conj(v)
+	assert outer(u, v) == conj(u)*v
+
+
+def test_apply_algebra_lowering_complex(self):
+	cell = triangle
+	element = FiniteElement("Lagrange", cell, 1)
+
+	v = TestFunction(element)
+	u = TrialFunction(element)
+
+	gv = grad(v)
+	gu = grad(u)
+	gush = gu.ufl_shape
+	gvsh = gv.ufl_shape
+	guinds = indices(len(gush))
+	gvinds = indices(len(gvsh))
+
+	a = dot(gu, gv)
+	b = inner(gu, gv)
+	c = outer(gu, gv)
+
+	lowered_a = apply_algebra_lowering(a)
+	lowered_b = apply_algebra_lowering(b)
+	lowered_c = apply_algebra_lowering(c)
+	lowered_a_index = lowered_a.index()
+	lowered_b_index = lowered_b.index()
+	lowered_c_indices = lowered_c.indices()
+
+	assert lowered_a == gu[lowered_a_index]*conj(gv[lowered_a_index])
+	assert lowered_b == gu[lowered_b_index]*conj(gv[lowered_b_index])
+	assert lowered_c == as_tensor(conj(gu[lowered_c_indices[0]])*gv[lowered_c_indices[1]], (lowered_c_indices[0],)+(lowered_c_indices[1],))
+
+
+def test_remove_complex_nodes(self):
+	cell = triangle
+	element = FiniteElement("Lagrange", cell, 1)
+
+	u = TrialFunction(element)
+	v = TestFunction(element)
+	f = Coefficient(element)
+
+	a = conj(v)
+	b = real(u)
+	c = imag(f)
+	d = conj(real(v))*imag(conj(u))
+
+	assert remove_complex_nodes(a) == v
+	assert remove_complex_nodes(b) == u
+	assert remove_complex_nodes(c) == f
```

Comment:
Yes, I was confused by this too.  I think it is because it "removes" complex nodes in the sense of stripping them from the dag and replacing them with their operand.